File: go/proverb/proverb.go
// Reproduce the for want of a nail proverb with the given set of things to be lost
package proverb

import "fmt"

// Reproduce the for want of a nail proverb with the given items.
// @param rhyme: Slice of string with the items to be lost (in-order)
// @returns: The computed proverb as a slice of string with one line per entry.
// Empty if no items and just the final line if only one item
func Proverb(rhyme []string) []string {
	var ret []string

	for i := 0; i < len(rhyme)-1; i++ {
		ret = append(ret, fmt.Sprintf("For want of a %s the %s was lost.", rhyme[i], rhyme[i+1]))
	}

	if len(rhyme) >= 1 {
		ret = append(ret, fmt.Sprintf("And all for the want of a %s.", rhyme[0]))
	}

	return ret
}

File: php/proverb/Proverb.php
<?php
// Recreate the for want of a nail, a kingdom was lost proverb with code.

declare(strict_types=1);

// Generate the for want of a nail, a kingdom was lost proverb with code.
class Proverb
{
    // Recreate the for want of a nail, a kingdom was lost proverb
    // with the pieces recieved as an argument.
    // @param $pieces: The objects lost in order.
    // @returns: Array of strings for each line of the proverb.
    public function recite($pieces)
    {
        $ret = array();

        if (count($pieces) >= 2) {
            for($i = 0; $i < count($pieces) - 1; $i++) {
                $ret[] = "For want of a {$pieces[$i]} the {$pieces[$i + 1]} was lost.";
            }
        }
        if (count($pieces) >= 1) {
            $ret[] = "And all for the want of a {$pieces[0]}.";
        }

        return $ret;
    }
}
